plugins/Synchro/SynchroSynth.cpp

```diff
@@ -0,0 +1,565 @@
+static inline float trih(float x, float harmonic)
+{
+	return tri(x) + ((tri(MAGIC_HARMONICS[0][0] * x) * MAGIC_HARMONICS[0][1]
+		+ tri(MAGIC_HARMONICS[1][0] * x) * MAGIC_HARMONICS[1][1])) * harmonic;
+}
+
+//Linear interpolation for envelopes
+//from: the initial value at x = 0
+//to: the final value at x = 1
+//x: the progress between the two values between 0 and 1
+static inline float lerp(float from, float to, float x)
+{
+	return from + x * (to - from);
+}
+
+//Exponential interpolation for envelopes
+//from: the initial value at x = 0
+//to: the final value at x = 1
+//x: the progress between the two values between 0 and 1
+static inline float expInterpol(float from, float to, float x)
+{
+	return from + fastSqrt(x) * (to - from);
+}
+
+//Detunes by octaves
+//pitch: the input pitch
+//detune: the amount of octaves to detune by
+static inline float DetuneOctaves(float pitch, float detune)
+{
+	return pitch * exp2(detune);
+}
+
+//Fast hyperbolic tangent approximation function
+//
+static inline float fastTanh(const float x)
+{
+	const double absX = fabs(x);
+	const double sqrX = x * x;
+	const double z = x * (0.773062670268356 + absX + (0.757118539838817 + 0.0139332362248817 * sqrX * sqrX) * sqrX * absX);
+	return z / (0.795956503022967 + fabs(z));
+}
+
+//Waveform function for the Synchro synthesizer.
+//x: input phase in radians (please keep it between 0 and 2PI)
+//drive: how much atan distortion is applied to the waveform
+//sync: hard sync with harmonic multiple
+//chop: how strong the amplitude falloff is per waveform period
+//harmonic: how strong the magic harmonics are
+static inline float SynchroWaveform(float x, float drive, float sync, float chop, float harmonic)
+{
+	return fastTanh(trih(F_PI_2 + x * sync, harmonic) * drive / powf(F_2PI / (F_2PI - x), chop));
+}
+
+//SynchroNote empty constructor
+SynchroNote::SynchroNote(NotePlayHandle * nph) :
+	nph(nph)
+{
+
+}
+
+void SynchroNote::nextStringSample(sampleFrame &outputSample, sample_rate_t sample_rate,
+	const float & modulationStrength, const float & modulationAmount, const float & harmonics,
+	const SynchroOscillatorSettings & carrier, const SynchroOscillatorSettings & modulator)
+{
+	float freqToSampleStep = F_2PI / sample_rate;
+	//How long (in frames) this note has been playing
+	float NoteTime = nph->totalFramesPlayed();
+
+	//Modulator is calculated first because it is used by the carrier
+	//Find position in modulator waveform
+	modulatorSampleIndex += freqToSampleStep * DetuneOctaves(nph->frequency(), modulator.Detune);
+	//Make sure phase is always between 0 and 2PI
+	while (modulatorSampleIndex >= F_2PI) { modulatorSampleIndex -= F_2PI; }
+
+	//Get modulator waveform at position
+	float modulatorSample = SynchroWaveform(modulatorSampleIndex,
+		modulator.Drive, modulator.Sync, modulator.Chop, harmonics) * (modulationStrength * modulationAmount);
+
+	//Modulator envelope
+	//Using sample_rate here gives an incorrect envelope length for some reason
```

Comment:
You're passing `Engine::mixer()->processingSampleRate() * SYNCHRO_OVERSAMPLE` here, but you're comparing with `NoteTime` which is based on `Engine::mixer()->processingSampleRate()`.

Reply:
A quick fix is to just make `NoteTime = nph->totalFramesPlayed() * SYNCHRO_OVERSAMPLE`, but I feel like this might be a band-aid on the underlying issue of how I'm handling envelopes in general. I'll come back to this.

Reply:
I wrote this comment to explain why you shouldn't use `sample_rate`(`= Engine::mixer()->processingSampleRate() * SYNCHRO_OVERSAMPLE`), not to say you're doing something wrong. I think you should revert the `noteTime` change to get the correct behavior.

Reply:
`nph->totalFramesPlayed()` will give you the number of frames based on `processingSampleRate()`, so you can compare with the envelope frames below without multiplying/dividing by `SYNCHRO_OVERSAMPLE`.

Reply:
Oh, I even found `noteTime` doesn't change during a rendering period because you're ignoring the frame offset in this function(`frame` in `playNote()`).

Reply:
I don't quite understand that last comment (there's still a lot I don't understand about audio programming). What's frame offset?